Fix Xcode Analyser warning by removing nullable (default) initializer

hansemannn opened the pull request:
Hey guys,

first of all, great utility library, love it! I use it to support clustering with Titanium and it works as expected. While fixing some recent Analyser warnings in my project, I noticed that the library also had one (trivial) warning. We use the following to disable the default initializer:
```objc
- (instancetype)init {
  self = [self initWithNullableMap:nil algorithm:nil renderer:nil];
  if (self) {
    [self doesNotRecognizeSelector:_cmd];
    self = nil;
  }
  return self;
}
```
The `initWithNullableMap:algorithm:renderer` method then assigns `nil` values to the properties and throws the `Unrecognized Selector` system error. But the `algorithm` property is non-null in the interface, so it throws the warning. Simple as that. The other ones are already `nullable`, so they pass the Analyser. With this PR, the behavior is normalized, so all cluster manager properties have the same nullability. 

An alternate solution would be to solve the disallowed `init` initializer by using the following:

```objc
- (id)init {
    [self release];
    @throw [NSException exceptionWithName:NSInternalInconsistencyException
                                   reason:@"The default initializer is not allowed. Please use initWithMap:algorithm:renderer instead!"
                                 userInfo:nil];
    return nil;
}
```

Let me know what you prefer. Thx and happy new year!

hansemannn commented:
`I signed it!` 😄

mountainvat commented:
Thanks for the analysis and PR. I think we prefer to throw on the init. The arguments are supposed to be non-nullable by the NS_ASSUME_NONNULL_BEGIN pragma.

Also can we use NSException raise instead of @throw?

Thanks.

mountainvat commented:
Sorry maybe I did not make it clear but we'd prefer to throw in the init method and leaving the algorithm property as non-nullable. It really should not be null for the cluster manger to be useful.

hansemannn commented:
@mountainvat Sorry for the late feedback, we should definitely use the `NSException` here. I'll update the PR asap!

hansemannn commented:
@mountainvat Updated the header to disallow the default init. We may want to remove the `initWithNullableMap:algorithm:renderer` as well and move the observer to the designated initializer (`initWithMap:algorithm:renderer`).

mountainvat commented on src/Clustering/GMUClusterManager.h:
```diff
@@ -45,13 +45,16 @@ NS_ASSUME_NONNULL_BEGIN
 @interface GMUClusterManager : NSObject<GMSMapViewDelegate>
 
+/** Disallow default initializer */
+- (id)init __attribute__((unavailable("Use initWithMap:algorithm:renderer: instead.")));
```
I think we'd like to use a macro like NS_UNAVAILABLE. Can you please find out from which SDK version was the macro introduced?

hansemannn replied:
I followed latest Apple API's, making it possible to print a message together with the error, so the dev will know which initializer to use, see [here](http://nshipster.com/__attribute__/). `NS_UNAVAILABLE` is fine, but not really helpful for the dev-workflow. You decide!

mountainvat commented on src/Clustering/GMUClusterManager.m:
```diff
@@ -50,16 +50,6 @@ @implementation GMUClusterManager {
 }
 
-// Disables init.
-- (instancetype)init {
```
Since we've removed default init method, can we also remove the initWithNullableMap and only keep 
```
- (instancetype)initWithMap:(GMSMapView *)mapView
                  algorithm:(id<GMUClusterAlgorithm>)algorithm
                   renderer:(id<GMUClusterRenderer>)renderer
```

hansemannn commented:
Ok, here we go! 
- Removed the `initWithNullableMap` initializer
- Used `NS_UNAVAILABLE` and `NS_DESIGNATED_INITIALIZER` to define a proper initializer usage
- Fixed some docs to unify the common documentation pattern

This makes fun! Thank you @mountainvat  for the guidance 🚀

mountainvat commented on src/Clustering/GMUClusterManager.h:
```diff
@@ -45,12 +50,19 @@ NS_ASSUME_NONNULL_BEGIN
+/**
+ * Returns a new instance of the GMUClusterManager class defined by it's |algorithm| and |renderer|.
+ */
+- (instancetype)initWithMap:(GMSMapView *)mapView algorithm:(id<GMUClusterAlgorithm>)algorithm renderer:(id<GMUClusterRenderer>)renderer NS_DESIGNATED_INITIALIZER;
```
Please wrap this as in the original declaration. This will allow lines to fit within the 100 character limit and assist readability.

hansemannn replied:
Sure thing. Just followed Apple's best practices, where they keep header declarations inline

mountainvat replied:
We are trying to keep the coding style to the same as the GoogleMaps headers. In general we want lines to be within 100 character limit for readability purposes (e.g code fits within 1 screen width).

Until we all have agreed to allow inline header declaration, we'd like to keep it this way. Thanks for the understanding.

mountainvat commented on src/Clustering/GMUClusterManager.h:
```diff
@@ -79,19 +91,26 @@ NS_ASSUME_NONNULL_BEGIN
  */
-- (void)setDelegate:(id<GMUClusterManagerDelegate> _Nullable)delegate
-        mapDelegate:(id<GMSMapViewDelegate> _Nullable)mapDelegate;
+- (void)setDelegate:(id<GMUClusterManagerDelegate> _Nullable)delegate mapDelegate:(id<GMSMapViewDelegate> _Nullable)mapDelegate;
```
Same here please revert this line.

mountainvat commented:
@hansemannn Thanks again for the contribution.
